### PR TITLE
Add recipe for libpnetcdf

### PR DESCRIPTION
**Pull request by xylar**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/libpnetcdf/meta.yaml
Comment (xylar) [inline recipes/libpnetcdf/meta.yaml:72]: @zklaus, would you be willing to be a maintainer of the PNetCDF library?  It seems like it would make sense since you're a maintainer (and a very helpful one!) for [parallelio](https://github.com/conda-forge/parallelio-feedstock).

File: recipes/libpnetcdf/meta.yaml
Comment (xylar) [inline recipes/libpnetcdf/meta.yaml:31]: We have had trouble with `run_exports` from netcdf-fortran in the `parallellio` feedstock, so I'm assuming that might be an issue here, too.
https://github.com/conda-forge/parallelio-feedstock/blob/main/recipe/meta.yaml#L33-L34

File: recipes/libpnetcdf/meta.yaml
Comment (zklaus) [reply recipes/libpnetcdf/meta.yaml:72]: Sure, can do :+1:

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:72]: Great, thank you!  Feel free to review and recommend changes.

**Comment by xylar** [comment]
@conda-forge/help-c-cpp, I'm sure you have a giant backlog but just wanting to check in about when you might be able to review this.

File: recipes/libpnetcdf/meta.yaml
Comment (carterbox) [inline recipes/libpnetcdf/meta.yaml:43]: You have overdepending/static linking warnings from hdf5. You either don't need hdf5 or you are static linking (which is not allowed).

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=646866&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=5165

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/libpnetcdf/meta.yaml
Comment (carterbox) [inline recipes/libpnetcdf/meta.yaml:57]: This recipe has a run_export which means there should be some existence tests for header files and package config files?

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/libpnetcdf/meta.yaml
Comment (xylar) [reply recipes/libpnetcdf/meta.yaml]: It looks like you're right and hdf5 isn't a dependency.  I don't know where I got the misimpression that it was.

File: recipes/libpnetcdf/meta.yaml
Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:55]: Again, good call.  I've added these.  Let's see how the testing goes.

**Comment by xylar** [comment]
@carterbox, thank you for your review and your patience with my inexperience at this.  This is ready for you to re-reveiw when you get a chance.

**Comment by xylar** [comment]
@conda-forge/help-c-cpp, ready for review!

File: recipes/libpnetcdf/meta.yaml
Comment (carterbox) [inline recipes/libpnetcdf/meta.yaml:43]: There is an overlinking warning on OSX for libnetcdf. Please check that you're not static linking.

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=662403&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=7450

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/libpnetcdf/meta.yaml
Comment (jedwards4b) [reply recipes/libpnetcdf/meta.yaml:43]: Is this the issue holding up this PR?  The build link is no longer there, can you re trigger?  And if possible give more detail as to the issue - the configure line above clearly specifies shared and not static libraries.

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:43]: Here is the same but for mpich instead of openmpi:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=722946&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=7598

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:43]: @carterbox and @jedwards4b, even with `libnetcdf` 4.9.1, which should be built only as a dynamic library, I am seeing:
```
WARNING (libpnetcdf): run-exports library package conda-forge::libnetcdf-4.9.1-mpi_openmpi_haee62c1_3 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```
I really don't understand why.  Any advice on how to fix this would be appreciated.

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:43]: @jedwards4b, is there any way that my recipe could be misconfigured (for OSX in particular) and not be linking `libpnetcdf` with `libnetcdf` at all?

Comment (jedwards4b) [reply recipes/libpnetcdf/meta.yaml:43]: Linking pnetcdf to netcdf is a new feature and one that I haven't used.  Do you need it?  What if we just remove the netcdf dependancy?

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:43]: I see, I assumed it was required.  We can remove it.

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml:43]: What about `netcdf-fortran`.  I guess it can't need that either if it doesn't need `netcdf-c`?

Comment (xylar) [reply recipes/libpnetcdf/meta.yaml]: @jedwards4b, I think I'm starting to understand.  I have `--with-netcdf4=${PREFIX}` in the build and I will need to remove that if we're not using `libnetcdf`.  Otherwise, I see (predictably):
```
      ------------------------------------------------------------
      Missing NetCDF-4 header files 'netcdf.h' or 'netcdf_meta.h'
      required to build PnetCDF with NetCDF-4 support. Use option
          --with-netcdf4=/path/to/implementation
      to specify the location of NetCDF-4 installation. In addition,
      please make sure the MPI C compiler is compatible with the one
      used to build NetCDF-4. Check 'config.log' for more information.
      Stopping ...
      ------------------------------------------------------------
```
Would that be your suggestion to begin with?  Build without the `--with-netcdf4` flag?

Assuming that's the case, I'll try removing it.

File: recipes/libpnetcdf/meta.yaml
Comment (xylar) [inline recipes/libpnetcdf/meta.yaml:71]: @jedwards4b, are you also interested in being a maintainer?  If so, your help would be quite valuable.

Comment (jedwards4b) [reply recipes/libpnetcdf/meta.yaml:71]: Sure, please add me as a maintainer.  Thanks

**Comment by xylar** [comment]
@conda-forge/help-c-cpp, ready for review!

**Comment by xylar** [comment]
@conda-forge/help-c-cpp and @carterbox, I have removed `libnetcdf` and `netcdf-fortran` as dependencies at least for now.  We can explore in the feedstock whether we can add `libnetcdf` back as a dependency without the overdependency warning.

**Comment by xylar** [comment]
Thanks a bunch @carterbox.  I know this was a slog.

**Pull request merged**
